File: src/main/java/algorithms/stacks/MaxHistogramArea.java
package algorithms.stacks;

import java.util.Scanner;
import java.util.Stack;

/**
 * Created by vinaysaini on 6/7/16.
 */

/**
 * Given an array of integers representing height of a histogram
 * Width is same as 1 unit - Find max Area
 * Complexity O(n)
 */
public class MaxHistogramArea {
    public static void main(String[] args) {
        Scanner in = new Scanner(System.in);
        int n = in.nextInt();
        int[] ar = new int[n];
        for (int i = 0; i < n; i++) {
            ar[i] = in.nextInt();
        }
        System.out.println(findMaxArea(ar));
    }

    static int findMaxArea(int[] ar) {
        Stack<Integer> stk = new Stack<>();
        int n = ar.length;
        int area = 0;
        for(int i = 0; i < n; i++) {
            if(stk.isEmpty()) stk.push(i);
            else {
                while(!stk.isEmpty() && ar[i] < ar[stk.peek()]) {
                    int idx = stk.pop();
                    int d = stk.isEmpty() ? i : i - stk.peek() - 1;
                    area = Math.max(area, ar[idx]*d);
                }
                stk.push(i);
            }
        }
        while(!stk.isEmpty()) {
            int idx = stk.pop();
            int d = stk.isEmpty() ? n : n - stk.peek() - 1;
            area = Math.max(area, ar[idx]*d);
        }
        return area;
    }
}

File: src/main/java/algorithms/stacks/MaxHistogramArea.cpp
#include <iostream>
#include <vector>
#include <stack>
#include <algorithm>
using namespace std;

int findMaxArea(vector<int> &ar) {
    stack<int> stk;
    int n = ar.size();
    int area = 0;
    int i = 0;
    while(i < n) {
        if(stk.empty() || ar[stk.top()] <= ar[i])
            stk.push(i++);
        else {
            int t = stk.top(); stk.pop();
            int width = stk.empty() ? i : i - stk.top() - 1;
            area = max(area, ar[t] * width);
        }
    }
    while(!stk.empty()){
        int t = stk.top(); stk.pop();
        int width = stk.empty() ? i : i - stk.top() - 1;
        area = max(area, ar[t] * width);
     }
    return area;
}

int main() {
    int n;
    cin >> n;
    vector<int> ar(n);
    for (int i = 0; i < n; i++) {
        cin >> ar[i];
    }
    cout << findMaxArea(ar) << endl;
    return 0;
}